include/pybind11/pytypes.h

```diff
@@ -298,6 +298,22 @@ class handle : public detail::object_api<handle> {
 
 #ifdef PYBIND11_HANDLE_REF_DEBUG
 private:
+    void throw_gilstate_error(const std::string &function_name) const {
+        fprintf(stderr,
+                "%s is being called while PyGILState_Check() is failing. "
+                "This usually indicates that you are calling pybind11 functions without holding "
+                "the GIL or before "
+                "Python (and/or this Python module) is finished initializing.\n",
+                function_name.c_str());
+        if (Py_TYPE(m_ptr)->tp_name != nullptr) {
+            fprintf(stderr,
+                    "The failing %s call was triggered on a %s object.\n",
+                    function_name.c_str(),
+                    Py_TYPE(m_ptr)->tp_name);
+        }
+        throw std::runtime_error(function_name + " PyGILState_Check() failure.");
```

Comment:
`stdout`/`stderr` buffering could make the output confusing (or even not show up at all if the buffers are never flushed before the process dies). I recommend adding

```
fflush(stdout);
... fprintf(stderr) ...
fflush(stderr);
```
and only then throw the exception.

Another thought, not sure about it myself because of the tedious technicalities involved & maybe loss of robustness: `throw std::runtime_error(same_message_as_sent_to_stderr);`

This might be better for some environments, such as long-running RPC servers with `stderr` streamed to some huge log file, while exceptions are reported in more outstanding ways (e.g. a typical situation at Google).

Reply:
Yeah, let me add a bunch of fflush calls. Let's not fiddle with exceptions.

Reply:
> Yeah, let me add a bunch of fflush calls. Let's not fiddle with exceptions.

Sounds good.

BTW/FWIW I had a similar struggle with some intermediate version of #1895, going back and forth, trying `std::terminate` instead of throwing an exception. It didn't really help on Windows, I couldn't find a way to make this perfect.

IOW what you have looks as good as it gets to me.